pkg/gadgets/trace/tcpdrop/tracer/bpf/tcpdrop.bpf.c

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: GPL-2.0
+//
+// Based on tcpdrop(8) from BCC
+//
+// Copyright 2018 Netflix, Inc.
+// 30-May-2018    Brendan Gregg   Created this.
+// 15-Jun-2022    Rong Tao        Add tracepoint:skb:kfree_skb
+// Copyright 2023 Microsoft Corporation
+
+#include <vmlinux/vmlinux.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+
+#include "tcpdrop.h"
+
+/* Define here, because there are conflicts with include files */
+#define AF_INET		2
+#define AF_INET6	10
+
+// we need this to make sure the compiler doesn't remove our struct
+const struct event *unusedevent __attribute__((unused));
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+// This struct is defined according to the following format file:
+// /sys/kernel/tracing/events/skb/kfree_skb/format
+struct kfree_skb_ctx {
+	/* The first 8 bytes is not allowed to read */
+	unsigned long pad;
+
+	void *skbaddr;
+	void *location;
+	unsigned short protocol;
+	int reason;
+};
+
+static __always_inline int __trace_tcp_drop(void *ctx, struct sock *sk, struct sk_buff *skb, int reason)
+{
+	if (sk == NULL)
+		return 0;
+
+	u_int8_t *tcphdr = (u_int8_t *)(BPF_CORE_READ(skb, head) + BPF_CORE_READ(skb, transport_header));
+	struct inet_sock *sockp = (struct inet_sock *)sk;
+	struct task_struct *task = (struct task_struct*) bpf_get_current_task();
+
+	struct event event = {};
+	event.timestamp = bpf_ktime_get_boot_ns();
+	event.af = BPF_CORE_READ(sk, __sk_common.skc_family);
+	event.state = BPF_CORE_READ(sk, __sk_common.skc_state);
+	event.reason = reason;
+	bpf_get_current_comm(&event.comm, sizeof(event.comm));
+	event.pid = bpf_get_current_pid_tgid() >> 32;
+	event.mntns_id = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+	bpf_probe_read_kernel(&event.tcpflags, sizeof(event.tcpflags), tcphdr+13);
+
+	BPF_CORE_READ_INTO(&event.dport, sk, __sk_common.skc_dport);
+	if (event.dport == 0)
+		return 0;
+
+	BPF_CORE_READ_INTO(&event.sport, sockp, inet_sport);
+	if (event.sport == 0)
+		return 0;
+
+	if (event.af == AF_INET) {
+		BPF_CORE_READ_INTO(&event.saddr_v4, sk, __sk_common.skc_rcv_saddr);
+		if (event.saddr_v4 == 0)
+			return 0;
+
+		BPF_CORE_READ_INTO(&event.daddr_v4, sk, __sk_common.skc_daddr);
+		if (event.daddr_v4 == 0)
+			return 0;
+
+	} else if (event.af == AF_INET6) {
+		bpf_probe_read_kernel(&event.daddr, sizeof(event.daddr),
```

Comment:
nit: `event.daddr, sk, __sk_common.skc_v6_daddr.in6_u.u6_addr32);` 
as in https://github.com/iovisor/bcc/blob/594ab9a9b0e3dfa7cf8fc7f28923ba0eb33a5c66/libbpf-tools/tcpconnlat.bpf.c#L82?